text/0000-optimize-by-default.md

```diff
@@ -0,0 +1,185 @@
+Rust.**
+
+## Why should we do this before 1.0?
+
+Because doing it after 1.0 breaks backwards compatibility on a command-line
+level. People will expect `rustc`/`cargo` to behave a certain way and will encode
+those assumptions into their bash command history/build files/scripts/whatever.
+
+# Detailed design
+
+Change `rustc` and `cargo` to produce optimized binaries by default.
+
+`cargo` currently has a `--release` flag; it should also get a `--debug` flag to
+trigger a debug build. By default, running `cargo build` is the same as running
+`cargo build --release`.
+
+Similarly, `rustc` should by default run with the same flags that it would
+receive from `cargo` if it were passed `--release`. It might make sense to also
+add support for `--release` & `--debug` flags to `rustc`, so that with
+`--debug` it becomes easy to instruct the compiler to not include optimizations
+and to include debug info.
+
+Debate is welcome on the specifics as long as "unadorned" `rustc` and `cargo
+build` calls produce optimized binaries.
+
+# Drawbacks
+
+## Slower builds by default
+
+Builds will only be slower _by default_; in other words, anyone calling `rustc
+foo.rs` today can trivially call `rustc -C optlevel=0 foo.rs` or `rustc --debug
+foo.rs` if the extra `rustc` flags are implemented.
+
+Note the failure state of debug-by-default: if you wanted an optimized build
+instead, your production binary is now incredibly slow. As someone who has
+actually pushed unoptimized binaries to production by accident, I can
+confidently say _this costs a lot of money._
+
+For opt-by-default, if you actually wanted a debug build the failure state is
+"you waited longer for your build to finish than you should have," which while
+annoying, has _far_ less potential for damage.
+
+## Breaks convention with compilers for other languages
+
+While this is a real concern, it's the result of those compilers continuing to
+carry a legacy burden. C compilers need to be backwards compatible with the
+state-of-the-art from many decades ago, and C++ compilers need to be compatible
+with C compilers. In other words, they don't have a choice.
+
+Rust has been breaking plenty of language conventions for the sake of both
+memory-safety and safety in general; it should continue on this commendable path
+without being hamstrung by decisions made in 1972.
+
+# Alternatives
+
+## Only have `cargo` echo a message stating `debug` or `release` build mode
+
+This has been recently talked about, and while it's a welcome and useful
+feature, it sadly isn't enough. A Rust newcomer who's only used Python, Ruby,
+JavaScript, Java, C# or a similar language is not going to understand the
+consequences of a "building in debug mode" message.
+
+Not to mention that it's easy to miss such messages when `cargo build` is run by
+a script that will also call plenty of other commands and will thus have lots of
+output.
+
+Not to mention it does nothing for `rustc` calls by newbies. See the next
```

Comment:
@tbu- "trying" to say something that's OK doesn't in any way justify saying it in a less-than-helpful way. I'm not commenting on whether this particular instance is OK or not, but, in general, responding to "please stay civil" with "they meant to say *X*" isn't productive.

---

> But you still end up with problems like banner blindess.

FWIW, banner blindness doesn't really apply to people using the tool for the first time, as  newbies who are accidentally benchmarking non-optimised code will be. Of course, this doesn't address the "accidentally deploy unopt binaries" case.

Although, on that note, Cargo now outputs such artefacts to `./target/debug/...` rather than just `./target/...`, so it does take more explicit effort to deploy those rather than the `./target/release/` ones as intended.